LoopKit/GlucoseKit/GlucoseCondition.swift

```diff
@@ -0,0 +1,75 @@
+//
+//  GlucoseCondition.swift
+//  LoopKit
+//
+//  Created by Darin Krauss on 9/3/21.
+//  Copyright © 2021 LoopKit Authors. All rights reserved.
+//
+
+import HealthKit
+
+public enum GlucoseCondition: Equatable, RawRepresentable {
```

Comment:
curious why GlucoseRangeCategory wasn't extended? Is there only a need to know when the range is exceeded? Is there no value to capture additional user context for the glucose value?

Reply:
I suppose I could have done that, but then I'd have to serialize all of that into Core Data, when I only need the aboveRange/belowRange and its threshold. I'll let @ps2 chime in here.

Reply:
GlucoseRangeCategory is a set of categories that are used for color coding glucose display based on Dexcom specific ranges, that are intended to convey severity to the user. These probably should've been isolated to dexcom somehow. 
What we need here is a CGM agnostic flag for describing measurements that were beyond the sensing capabilities of the sensor. So I think a separate enum makes sense.

I do wonder if the HKQuantity value being stored here is redundant with the HKQuantity stored in the glucose reading itself, though.

Reply:
What does the threshold mean, if it's not the maximum value that can be reported?

Reply:
I.e. I would interpret an entry in the database with a value of 401 and a condition of "aboveRange" to mean that the sensor is reporting bg is >= 401.

Reply:
I don't think we should be storing metadata about sensors/glucometers here (a display threshold would be such a piece of data). This is a storage of glucose sensor readings; any metadata should be about the reading itself.

Reply:
Okay, the trick here is that the Tidepool data model uses two different properties when reporting an out-of-range glucose value. It looks something like this:
```
    "value": 401,
    "annotations": {
        "code": "bg/out-of-range",
        "value": "high",
        "threshold": 400
    }
```
So, there is the value and the threshold (above which are invalid.

Reply:
(And no, the Tidepool data model is not going to change for this one. It has been this way for years and it is used in several places.)

Reply:
So in the Tidepool backend model, what would a `value` of 450 and `threshold` of 400 mean? How would it be used? Or, even more interestingly, a `value` of 400 and `threshold` of 450?

Reply:
The latter would be an error.

As to the former, I don't have an easy answer here. I'm not an expert on the many, many devices the Tidepool Uploader supports and I was not involved in the data model design of the out-of-range annotations (as it was before my time, I believe). However, this is the data model design and changing it would cause significant work in Tidepool Uploader and the Tidepool backend (to migrate all older data). So, I don't expect this to change.

Now, as to how Tidepool Loop (through Tidepool Service) supports that, I can think of a couple of options: 
- Do something like what I already implemented where the threshold value (e.g. 400) is specified along with the out-of-range value (e.g. 401).
- Do not include the threshold value and just note that the value is above or below range. Then Tidepool Service would just "know" that we have to add or subtract "1 unit" from the out-of-range value to find the threshold. (I guess "1 unit" would always be 1 mg/dL here since the GlucoseStore is unaware of the actual units being generated by devices or entered by users and always extracts it out of the HKQuantity as mg/dL.)

Thoughts?

Reply:
Yeah, I wasn't thinking we'd change the Tidepool backend model. I was wanting to understand the semantics so we can build a reasonable mapping. If we have our semantics such that value: 401, condition: aboveRange means that the estimated glucose is 401 or greater and has hit the boundaries of reportable values, that seems simple and not redundant, and from what you offer in the second option, it sounds like we can map that to the semantics of the Tidepool backend, though I must admit I still don't feel like I understand it completely.

Reply:
> GlucoseRangeCategory is a set of categories that are used for color coding glucose display based on Dexcom specific ranges, that are intended to convey severity to the user. These probably should've been isolated to dexcom somehow.

@ps2 `GlucoseRangeCategory` isn't intended just for Dexcom. Design of the CGM pill reports all CGM values like this, hence why a manual glucose reading also gets a `GlucoseRangeCategory`. It is used for color coding, but also other representations (e.g., HIGH or LOW). It is intended to be a CGM agnostic context based on user settings and CGM sensing capabilities.

Reply:
I think there are two issues here:

1) Where/how to report the out of range condition including how to persist it along with the glucose sample.

IMHO, the `GlucoseRangeCategory` is **not** a part of the glucose sample. It is instead the product of a glucose sample and a set of alert conditions (i.e. urgent low, low, and high levels) at the time of the glucose sample. It should not be stored with the glucose sample in CoreData nor should it be included in the uploading data to a remote data service.

2) What the out of range condition looks like.

Frankly, stored 39 and 401 as the value of the glucose sample is misleading. The 39 and 401 reported by Dexcom are effectively out of range indicators (flags for LO and HI). They do not represent actual glucose readings. (The actual reading might be much lower than 39 or much higher that 400, but Dexcom does not report that information, AFAIK.) So, in my mind, it would be more consistent to **not** include a value with the glucose sample, but rather report out of range LO or HI.

Furthermore, the 39 and 401 really represent that the value is either **below 40** or **above 400**. If Dexcom (or another device, perhaps more importantly) happened to report values as real numbers rather than integers then 39 and 401 wouldn't work.)

Now given that, I don't expect Loop to stop storing 39 and 401 (even though they aren't actual glucose values) because its been doing that for a while now. However, using the 39 and 401 value as an out of range indicator won't work as other devices may have other values that are considered out of range. Thus, a separate out of range indicator is necessary. 

Furthermore, since 39 and 401 are just placeholders for out of range, we really should be keeping track of the last value valid before going out of range. While this threshold isn't required by Loop/LoopKit, the Tidepool backend data model does expect it. Now, we could see if dropping the threshold value from the reported data would cause a problem, if there is too much resistance to capturing the threshold internally in Loop/LoopKit.

Thoughts?

Reply:
Let's get semantics clear first, and then decide on representation. The semantics of Dexcom reporting a 401 value is simply that the reading is higher than 400, right? It *is* a record of glucose measurement (it represents a sensor measurement of glucose at a specific time), but there is no specific glucose value; just a range (400-inf).  For these "out of range" readings, we want to track the range limits, since they may vary from sensor to sensor.

Using a condition field like you have added in this PR, that indicates above or below range, and using the existing value field to represent the bottom or top part of the range (depending on the condition), is one way to represent those semantics. 

You could certainly add another value field to represent the single needed range boundary, but now we have two measurement fields and a larger number of non-sensical values possible, like the ones we were discussing above (bg 401, limit 450). The smaller possible state space seems less open to wrong interpretations.

Using your point about 401 being misleading as the range boundary if considering other units, I agree. So we would store 400 as the range limit (non-inclusive), and store "aboveMeasurementRange" as the condition.

I think as long as we have an unambiguous mapping of our representation in core data to some clear semantics, then figuring out how to represent those semantics in other storage places (like HK or Tidepool backend) is a separate problem. HK doesn't have storage for a "Range" as a sample, so if we want to store these in HK (and I think we do), it will be a compromise. Probably doing what we do now, storing 401 and 39, with a metadata annotation is the best we can do.

Reply:
Okay, just so I understand what you are suggesting, is it:

1) Get rid of the threshold? Or,

2) Store 40/400 in the CachedGlucoseObject along with the condition (above/below range) and then if/when we translate to/from HealthKit we change the value from 40/400 to 39/401 and back along with a metadata annotation? Or,

3) Something else?

Reply:
I think #2. The HK representation isn't going to be great, just because there isn't an idiomatic way to store a range in HK. We'll probably want two metadata fields there: the condition and the (noninclusive) range limit. The glucose value would be the first integer value in mg/dL outside of the range; basically reproducing what Dexcom does, but also providing a consistent representation for other sensors that might have different limits.

Reply:
On a whim, I setup the Dexcom G6 app and the emulator with a 401 glucose value. After 3 hours, the Dexcom G6 app writes the 401 mg/dL as the Blood Glucose value, with a metadata field "Status" with value "HIGH".

BTW, it also includes the following metadata:
"Trend Rate": 12.7
"Trend Arrow": "None"
"Device Name": "10386270000221"
"Transmitter Time": "(UTC Time)"
"Time Zone": "Pacific Daylight Time" <<===

Reply:
Yeah, we could match some of that, since we don't have anything established for those. Not sure about Time Zone, as we'd want it to match the timezone in effect during the *reading*, not the time of storing to HK, and we don't store that in our local store.

Reply:
Okay, I'm a bit stuck on implementing your recommended changes. Specifically, handling the `CGMStatusIndicator.glucoseRangeCategory(for glucose: GlucoseSampleValue) -> GlucoseRangeCategory?` for Dexcom. If we restrict glucose quantity values to 40-400, then we have no way to know whether the 40 or 400 are valid or out-of-range when calculating the `GlucoseRangeCategory` (specifically the belowRange and aboveRange cases). The only solution I can see would be to add the `GlucoseCondition` to the `GlucoseSampleValue` type, but then that leads to a bunch of other changes particular to the Dexcom plugin (e.g `CGMReading`).

Reply:
I think a large part of the problem is that there is no clear dividing line where the value are converted from 401 (Dexcom) to 400 (Loop). There are (effectively) raw glucose samples sent back into Dexcom that need to be converted back into Dexcom space (401). I think this back-and-forth conversion is likely to just lead to all sorts of errors and confusion.

Reply:
Yeah, that API is probably not right. Leaking a sigil (401/39) out of the dexcom plugin is bad in the first place, and then asking Dexcom plugin to convert it into a semantic type is weird. We should have a clean semantic type to represent that reading when it leaves Dexcom. I think this is worth cleaning up, but I do see how it's expanding the amount of work you're having to tackle; maybe there's a way to break this work into smaller pieces?

Reply:
Should be all resolved now. 401 and 39 are no longer used anywhere. Any value below 40 or above 400 in Dexcom are converted to 40 with `.belowRange` condition and 400 with `.aboveRange` condition, respectively.